### PR TITLE
[Server] Start implementing lizmap server side

### PR DESCRIPTION
**Pull request by rldhont**
First tool is checking project lizmap acl configuration

### CODE REVIEW

**Comment by rldhont** [comment]
@Gustry @dmarteau @mdouchin feedback and review are welcomes even if it is a draft

File: server/lizmap_filter.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+        handler = self.iface.requestHandler()
+
+        groups = handler.requestHeader('X-Lizmap-User-Groups').split(',')
+        groups = [g.strip() for g in groups]
+        for g in groups:
+            if g in cfg_acl:
+                return
+
+        raise QgsOgcServiceException('Forbidden', 'No ACL permissions', responseCode=403)
```
Comment (dmarteau) [inline server/lizmap_filter.py]: That's will not work, afaik python exception will not be translated to C++ exception. You have to set the return code using the request handler.

Comment (rldhont) [reply server/lizmap_filter.py]: can I use ?
```
handler.setServiceException(QgsOgcServiceException('Forbidden', 'No ACL permissions', responseCode=403))
handler.setStatusCode(403)
```

File: __init__.py [hunk 2]
```diff
@@ -50,3 +50,13 @@
 def classFactory(iface):
     from .lizmap import Lizmap
     return Lizmap(iface)
+
+# Server Plugin only
+def serverClassFactory(serverIface):  # pylint: disable=invalid-name
+    """Load atlasprint class from file atlasprint.
```
Comment (Gustry) [inline __init__.py]:
```suggestion
    """Load lizmap server class.
```

File: server/lizmap_filter.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import os
+import json
+
+from qgis.core import Qgis, QgsMessageLog
+from qgis.server import QgsServerInterface, QgsOgcServiceException
```
Comment (Gustry) [inline server/lizmap_filter.py]:
```suggestion
from qgis.server import QgsServerInterface, QgsOgcServiceException, QgsServerFilter
```

File: server/lizmap_filter.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+from qgis.server import QgsServerFilter
+from qgis.core import Qgis, QgsMessageLog
```
Comment (Gustry) [inline server/lizmap_filter.py]:
```suggestion

```

File: server/lizmap_filter.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
+import os
+import json
+
+from qgis.core import Qgis, QgsMessageLog
+from qgis.server import QgsServerInterface, QgsOgcServiceException
+
+from qgis.server import QgsServerFilter
```
Comment (Gustry) [inline server/lizmap_filter.py]:
```suggestion

```

File: server/lizmap_filter.py (hunk 1 above)
Comment (dmarteau) [reply server/lizmap_filter.py]: If you use the first one, you do not need to use `setStastusCode`. If you use `setStatusCode` return a body error message.  

`setServiceException` may ensure that the error message will be properly formated.

File: server/__init__.py [hunk 6]
```diff
@@ -0,0 +1 @@
+# -*- coding: utf-8 -*-
```
Comment (Gustry) [inline server/__init__.py]: Not needed anymore

File: server/lizmap_filter.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
+import os
+import json
+import traceback
+
+from qgis.core import Qgis, QgsMessageLog
+from qgis.server import QgsServerInterface, QgsServerException, QgsOgcServiceException, QgsServerFilter
+
+from PyQt5.QtCore import QByteArray
+from PyQt5.QtXml import QDomDocument
```
Comment (Gustry) [inline server/lizmap_filter.py]: change to `qgis.PyQt`

File: server/lizmap_service.py [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+            exp = QgsExpression(e)
+            exp.setGeomCalculator(da)
+            exp.setDistanceUnits(project.distanceUnits())
+            exp.setAreaUnits(project.areaUnits())
+
+            if exp.hasParserError():
+                exp_parser_errors.append('Error "{}": {}'.format(e, exp.parserErrorString()))
+                continue
+
+            if not exp.isValid():
```
Comment (Gustry) [inline server/lizmap_service.py]: `hasEvalError`?

Comment (rldhont) [reply server/lizmap_service.py]: not yet, the expression as not been yet evaluated.

**Comment by Gustry** [comment]
@rldhont Alright, you can pull your branch and start working again on it. I have extracted all what I need. This branch is up to date with master.

**Comment by rldhont** [comment]
@Gustry how to make test ?

**Comment by Gustry** [comment]
`make test_server` from root folder

**Comment by rldhont** [comment]
ok thanks

**Comment by Gustry** [comment]
@dmarteau Do you have any ideas why the Python version used **inside** the container is different when we run on laptop or travis?

On @rldhont's laptop and mine : make tests gave us:

```
docker run --rm --name qgis-server-lizmap-test-3.4-67dfeb2 -w /src/test/server \
        -u 1000 \
        -v /home/etienne/dev/lizmap/lizmap-plugin:/src \
        -v $(pwd)/.local:/.local \
        -v /home/etienne/dev/lizmap/lizmap-plugin/.cache:/.cache \
        -e PIP_CACHE_DIR=/.cache \
        -e PYTEST_ADDOPTS="" \
        qgis-platform:3.4 ./run-tests.sh
Installing required packages...
============================= test session starts ==============================
platform linux -- Python 3.5.3, pytest-5.3.5, py-1.8.1, pluggy-0.13.1 -- /usr/bin/python3
```
> Python 3.5.3

While on travis, it gives:

```
$ docker run --rm --name qgis-server -u $(id -u) -v ${TRAVIS_BUILD_DIR}/.cache:/.cache -e PIP_CACHE_DIR=/.cache -w /src/test/server -e PYTEST_ADDOPTS="" -v ${TRAVIS_BUILD_DIR}:/src -v ${TRAVIS_BUILD_DIR}/.local:/.local 3liz/qgis-platform:3.4 ./run-tests.sh

Installing required packages...

============================= test session starts ==============================

platform linux -- Python 3.7.3, pytest-5.3.5, py-1.8.1, pluggy-0.13.1 -- /usr/bin/python3
```
> Python 3.7.3

as we can see on this line : https://travis-ci.org/3liz/lizmap-plugin/jobs/660255929#L215

How can we update to the Python version? 3.5 is not so good...

**Comment by dmarteau** [comment]
@Gustry It dépends on the image, earlier images are based on debian:stretch-slim, which use python 3.5. 
Take care that you may have to prefix the image name by '3liz/'. Try using `REGISTRY_URL=3liz` with your make command.

**Comment by rldhont** [comment]
Thanks, my tests failed ;-)

**Comment by rldhont** [comment]
I have found the issue in my tests ;-)

File: lizmap/server/core.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+class ServiceError(Exception):
+
+    def __init__(self, code: str, msg: str, responseCode: int = 500) -> None:
+        super().__init__(msg)
+        self.service = 'Lizmap'
+        self.msg = msg
+        self.code = code
+        self.responseCode = responseCode
+        QgsMessageLog.logMessage("%s request error %s: %s" % (self.service, code, msg), "lizmap", Qgis.Critical)
```
Comment (Gustry) [inline lizmap/server/core.py]: @dmarteau Can we switch to using logging? Or is-it a blocker for py-qgis-server?
* shorter to write
* The name of the logger (`lizmap` is already setup for us)
* More debug levels than QGIS Debug levels
* The logger has more connectors

Same line, by switching to `format` at the same time
```python
LOGGER.critical("{} request error {}: {}".format(self.service, code, msg))
```

Comment (dmarteau) [reply lizmap/server/core.py]: There is no point to use logging if your code is executed in the Qgis context - i.e as a server plugin: QgsMessageLog is appropriate in this case.  

If you use logging your logs wil no be output correctly with another wrapper implementation of qgis (fcgi for exemple) server. 

Also, with py-qgis-server you bypass the logging configuration of the server, which is not good.

Comment (Gustry) [reply lizmap/server/core.py]: Ok, let like this @rldhont. I will make a small wrapper later.

**Pull request merged**